### PR TITLE
Generalized RDF-star

### PR DESCRIPTION
**Pull request by pchampin**
This PR mostly extends the generalized parsers to support RDF-star.

In the process, I also updated the documentation, and some internal method, of the other parsers, to align with the latest terminology of RDF-star ("RDF-star" instead of "RDF*", "quoted triples" instead of "embedded triples").

### CODE REVIEW

**Comment by pchampin** [comment]
I forgot to mention a few other changes:
- I have added a 2nd generalized parser: Generalized N-Quads
- some formatting changes in multiple files have been caused by running the latest `cargo fmt`

Finally, I ran `cargo clippy` and fixed all the warnings... but then clippy fails with a compiler panic -- as can be seen in the CI... :-/

File: api/src/generalized.rs [hunk 1]
```diff
@@ -56,7 +57,7 @@ pub mod model {
         BlankNode(BlankNode<'a>),
         Literal(Literal<'a>),
         Variable(Variable<'a>),
-        Triple(&'a Triple<'a>),
+        Triple(&'a [GeneralizedTerm<'a>; 3]),
```
Comment (Tpt) [inline api/src/generalized.rs:60]: A specific struct or `Triple { subject: &GeneralizedTerm<'a>, predicate: &GeneralizedTerm<'a>, object: &GeneralizedTerm<'a> }` might be nicer but this is fine imho

File: turtle/src/triple_allocator.rs [hunk 2]
```diff
@@ -31,6 +31,14 @@ impl TripleAllocator {
         }
     }
 
+    pub fn complete_len(&self) -> usize {
+        self.complete_len
+    }
+
+    pub fn incomplete_len(&self) -> usize {
+        self.complete_len
```
Comment (Tpt) [inline turtle/src/triple_allocator.rs:39]: Maybe `incomplete_len`?

File: turtle/src/gtriple_allocator.rs [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+            string_stack: StringBufferStack::with_capacity(4),
+        }
+    }
+
+    pub fn complete_len(&self) -> usize {
+        self.complete_len
+    }
+
+    pub fn incomplete_len(&self) -> usize {
+        self.complete_len
```
Comment (Tpt) [inline turtle/src/gtriple_allocator.rs:39]: Maybe`incomplete_len`?

File: turtle/src/gnquads.rs [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+use crate::gtriple_allocator::GeneralizedTripleAllocator;
+use crate::ntriples::{parse_literal, skip_until_eol, skip_whitespace};
+use crate::shared::*;
+use crate::utils::*;
+use oxiri::IriRef;
+use rio_api::model::*;
+use rio_api::parser::*;
+use std::io::BufRead;
+
+/// A [TriG](https://www.w3.org/TR/trig/) streaming parser parsing generalized quads.
```
Comment (Tpt) [inline turtle/src/gnquads.rs:13]: TriG -> NQuads

File: turtle/src/gnquads.rs [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+///
+/// Warning: RDF-star is not supported yet.
```
Comment (Tpt) [inline turtle/src/gnquads.rs:15]: Seems to be

File: turtle/src/gnquads.rs [hunk 6]
```diff
@@ -0,0 +1,326 @@
+//! Implementation of a generalized RDF / RDF-star version of the Trig syntax
```
Comment (Tpt) [inline turtle/src/gnquads.rs:1]: TriG -> NQuads

File: turtle/src/gtrig.rs [hunk 7]
```diff
@@ -9,7 +11,7 @@ use rio_api::model::*;
 use rio_api::parser::GeneralizedQuadsParser;
 use std::collections::HashMap;
 use std::io::BufRead;
-use std::str;
+use std::mem::swap;
 
 /// A [TriG](https://www.w3.org/TR/trig/) streaming parser parsing generalized quads.
 ///
```
Comment (Tpt) [inline turtle/src/gtrig.rs:17]: About the line just below: RDF-star is now supported.

**Review by Tpt — COMMENTED** [review]
Looks great to me! Thank you!

I found a few typos (not very important...)

About the CI failure, I have opened a bug in the Clippy repo about it: https://github.com/rust-lang/rust-clippy/issues/10021

File: turtle/src/triple_allocator.rs (hunk 2 above)
Comment (pchampin) [reply turtle/src/triple_allocator.rs:39]: ouch! and I'm using this in multiple `debug_assert` and that didn't break ?!?

**Comment by Tpt** [comment]
> About the CI failure, I have opened a bug in the Clippy repo about it: https://github.com/rust-lang/rust-clippy/issues/10021

This has been fixed in the latest nightly build

File: api/src/generalized.rs (hunk 1 above)
Comment (pchampin) [reply api/src/generalized.rs:60]: I would rather keep it that way, as I really on being able to index into triples using `usize`s.

File: turtle/src/triple_allocator.rs (hunk 2 above)
Comment (pchampin) [reply turtle/src/triple_allocator.rs:39]: it didn't break because my assertions were wrong! But I now understand why, and I fixed them :)

**Comment by pchampin** [comment]
@Tpt thanks for these comments. I addressed them all (except for replacing `[GeneralizedTerm; 3]` with a custom `GeneralizedTriple` type, which was not a strong requirement IIUC.

**Pull request merged**